exchanges/gateio/gateio_wrapper.go

```diff
@@ -2578,3 +2559,215 @@ func (g *Gateio) GetCurrencyTradeURL(_ context.Context, a asset.Item, cp currenc
 		return "", fmt.Errorf("%w %v", asset.ErrNotSupported, a)
 	}
 }
+
+// WebsocketSubmitOrder submits an order to the exchange through the websocket
+// connection.
+// NOTE: Regarding spot orders, fee is applied to purchased currency.
+func (g *Gateio) WebsocketSubmitOrder(ctx context.Context, s *order.Submit) (*order.SubmitResponse, error) {
+	err := s.Validate(g.GetTradingRequirements())
+	if err != nil {
+		return nil, err
+	}
+
+	s.Pair, err = g.FormatExchangeCurrency(s.Pair, s.AssetType)
+	if err != nil {
+		return nil, err
+	}
+	s.Pair = s.Pair.Upper()
+
+	switch s.AssetType {
+	case asset.Spot:
+		req, err := g.getSpotOrderRequest(s)
+		if err != nil {
+			return nil, err
+		}
+
+		got, err := g.WebsocketSpotSubmitOrder(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+
+		sr, err := g.deriveSpotWebsocketOrderResponses(got)
+		if err != nil {
+			return nil, err
+		}
+		return sr[0], nil
+	case asset.Futures:
+		amountWithDirection, err := getFutureOrderSize(s)
+		if err != nil {
+			return nil, err
+		}
+
+		timeInForce, err := getTimeInForce(s)
+		if err != nil {
+			return nil, err
+		}
+
+		got, err := g.WebsocketFuturesSubmitOrder(ctx, &ContractOrderCreateParams{
+			Contract:    s.Pair,
+			Size:        amountWithDirection,
+			Price:       strconv.FormatFloat(s.Price, 'f', -1, 64),
+			ReduceOnly:  s.ReduceOnly,
+			TimeInForce: timeInForce,
+			Text:        s.ClientOrderID,
+		})
+		if err != nil {
+			return nil, err
+		}
+		sr, err := g.deriveFuturesWebsocketOrderResponses(got)
+		if err != nil {
+			return nil, err
+		}
+		return sr[0], nil
+	default:
+		return nil, common.ErrNotYetImplemented
+	}
+}
+
+// deriveSpotWebsocketOrderResponses returns the order submission responses for spot
+func (g *Gateio) deriveSpotWebsocketOrderResponses(responses []WebsocketOrderResponse) ([]*order.SubmitResponse, error) {
```

Comment:
I'm not clear why we're receiving and returning a slice when we're going to ignore `[1:]` on lines L2621 and 2594.

Seems to me like it should be deriveResponse, accepting a singe OrderResponse, and we should pass got[0] into it, and probably error on len(got) > 1.
Am I missing something?

Reply:
It was due to batch orders which I had together in a branch which I have split that out into another PR. I will do the suggestion of having a function for `deriveResponse`